### ISSUE TITLE
Bitsandbytes dos not support ROCm

### ISSUE THREAD

**Issue by luanrivello** [nixified-ai/flake#56]
While trying to run `nix run .#textgen-amd` I got the following error:

error:
       … while evaluating the attribute 'optionalValue.value'

         at /nix/store/xjviahzwa7x51vl51kc3c1k1n1jmhpd5-source/lib/modules.nix:854:5:

          853|
          854|     optionalValue =
             |     ^
          855|       if isDefined then { value = mergedValue; }

       … while evaluating a branch condition

         at /nix/store/xjviahzwa7x51vl51kc3c1k1n1jmhpd5-source/lib/modules.nix:855:7:

          854|     optionalValue =
          855|       if isDefined then { value = mergedValue; }
             |       ^
          856|       else {};

       (stack trace truncated; use '--show-trace' to show the full trace)

       error:

       text-generation-webui is not supported on AMD yet, as bitsandbytes does not support ROCm.

Also may I ask why not keep both KoboldAI and textgen as a part of the project?

**Comment by Airradda** [comment]
ROCm is not officially supported by bitsandbytes, see [#107](https://github.com/TimDettmers/bitsandbytes/issues/107) [#681](https://github.com/TimDettmers/bitsandbytes/issues/681), so for now, it's set to [throw an error](https://github.com/nixified-ai/flake/blob/7b9730e0a75bcd6375b0780aef75e406ea20a9a4/projects/textgen/default.nix#L50) when you try to use `nix run .#textgen-amd`. However, there is PR [#756](https://github.com/TimDettmers/bitsandbytes/pull/756) for adding support. I have managed to get the webui up and running just fine by swapping with the version in the PR, but run into missing module errors when trying to use any the model loaders. I am unsure if this is an issue related to swapping bitsandbytes or something else entirely. If, however, I get it working I will let you know and maybe make a PR if Nixified is okay with using an update to date [fork](https://github.com/arlo-phoenix/bitsandbytes-rocm-5.6).

**Comment by MatthewCroughan** [comment]
@Airradda yes, totally up for using a fork, maybe Nixpkgs is also okay with it, but I'll accept it as soon as you make it. For me, this repo is about making things work, regardless of hacks, and then eventually upstreaming as much as possible to Nixpkgs.

**Comment by Airradda** [comment]
> @Airradda yes, totally up for using a fork, maybe Nixpkgs is also okay with it, but I'll accept it as soon as you make it. For me, this repo is about making things work, regardless of hacks, and then eventually upstreaming as much as possible to Nixpkgs.

I'll start working on it after I figure out how the new structure is setup after the refactor (079f83b). My working copy simply overwrote bitsandbytes as I only have an AMD GPU and didn't need the Nvidia stuff, which obviously won't work here. 

It would probably be a good idea to have some info for contributing along with any relevant info for #2 and #26.

**Comment by MatthewCroughan** [comment]
@devloic

> since most of them use around 20GB.

The reason they use 20GB is because of the poorly defined and seemingly unlimited runtime dependencies. text-generation-webui ambiguously requires the kitchen sink a runtime, and if we want to make the functionality equivalent to what you just posted, we also have to provide the kitchen sink. We also have to reverse engineer what parts the kitchen sink is made of, because not all of the dependencies are cleanly described in their entirety in any readable file or document in the repo, such as the requirements.txt.

If we obtained equivalent functionality, it's probably true that we'd approach that size too. I don't think text-generation-webui is worth that amount of effort though, as [something written in Rust compatible with WGPU/BURN is hopefully going to be written in the future](https://github.com/Gadersd/llama2-burn/issues), which means all of these problems will disappear, as it's created on better foundations.

I will accept any PR that fixes text-generation-webui's quirks though.

**Comment by Airradda** [comment]
For me the instant gain is not having to have 2+ copies of ROCm and some of the other requirements as I have Text-Gen (docker), SD-WebUI(nix-stable-diffusion), and InvokeAI(nixified-ai) installed, having multiple copies of models as not everything is playing nice with symlinks (docker), and basically just not having to deal with docker.

Tackling extensions and their requirements (the kitchen sink) is going to be a whole nother mess.
